base/src/main/java/proguard/classfile/util/inject/CodeInjector.java

```diff
@@ -0,0 +1,216 @@
+            + ".";
+
+    assert injectStrategy != null
+        : "The injection location hasn't been provided. please use `.at(...)` to indicate the place to inject "
+            + renderInjectionContent(content.clazz, content.method, arguments)
+            + " into the target method.";
+
+    targets.forEach(
+        target -> {
+          CodeAttributeEditor editor = new CodeAttributeEditor();
+          InstructionSequenceBuilder code =
+              new InstructionSequenceBuilder((ProgramClass) target.clazz);
+
+          arguments.forEach(
+              argument ->
+                  code.pushPrimitiveOrString(argument.getValue(), argument.getInternalType()));
+          code.invokestatic(content.clazz, content.method);
+
+          target.method.accept(
+              target.clazz,
+              new AllAttributeVisitor(
+                  new AttributeNameFilter(
+                      Attribute.CODE,
+                      new AttributeVisitor() {
+                        @Override
+                        public void visitCodeAttribute(
+                            Clazz clazz, Method method, CodeAttribute codeAttribute) {
+                          editor.reset(codeAttribute.u4codeLength);
+                          InjectStrategy.InjectLocation[] injectLocations =
+                              injectStrategy.getAllSuitableInjectionLocation(
+                                  (ProgramClass) clazz, (ProgramMethod) method);
+                          for (InjectStrategy.InjectLocation location : injectLocations) {
+                            final BiConsumer<Integer, Instruction[]> inserter =
+                                location.shouldInjectBefore()
+                                    ? editor::insertBeforeOffset
+                                    : editor::insertAfterInstruction;
+
+                            inserter.accept(location.getOffset(), code.instructions());
+                          }
+                          codeAttribute.accept(clazz, method, editor);
+                        }
+                      })));
+        });
+  }
+
+  // Internal utility methods
+  private static String renderMethodSignature(Clazz clazz, Method method) {
+    return ClassUtil.externalFullMethodDescription(
+        clazz.getName(),
+        method.getAccessFlags(),
+        method.getName(clazz),
+        method.getDescriptor(clazz));
+  }
+
+  private static String renderInjectionContent(
+      Clazz clazz, Method method, List<InjectedArgument> arguments) {
+    return new StringBuilder()
+        .append(clazz.getName())
+        .append(method.getName(clazz))
+        .append("(")
+        .append(arguments.stream().map(Object::toString).collect(Collectors.joining(",")))
+        .append("):")
+        .append(ClassUtil.externalMethodReturnType(method.getDescriptor(clazz)))
+        .toString();
+  }
+
+  // Internal utility classes
+  private static class ClassMethodPair {
```

Comment:
There's a Pair class in MemberViewModel already, apparently. maybe better to extract and reuse that one